### PR TITLE
⚠️ Plugin phase 1.5 implementation

### PR DESCRIPTION
**Pull request by Adirio**
[Design doc](https://github.com/kubernetes-sigs/kubebuilder/blob/2b464030b6e3f52756e3a1f4c01369ebd58e6dcf/designs/extensible-cli-and-scaffolding-plugins-phase-1-5.md)

It would solve the following issues.
- Closes: #1543
- Closes: #1941
- Closes: #1942

**NOTE**: 
To ensure that all changes proposed here also works with the consumer (SDK), operator-framework/operator-sdk#4581 was created.

### CODE REVIEW

**Comment by camilamacedo86** [comment]
To ensure that all changes proposed here also works with the consumer without breaking changes and indeed make the things easier maintained in SDK, @Adirio creates https://github.com/operator-framework/operator-sdk/pull/4581.

My POV for this PR/changes: 

- We have been discussing a way to allow all these possibilities for months and @Adirio raised possible solutions and prove that is feasible with these changes. 
- It mainly covers all requirements for 2.0
- I am satisfied by only ensuring that now we allow all these possibilities without cause breaking changes that would affect the SDK users. 

So it shows we are able to reach a big milestone and regarding the code changes and the details I think we can indeed improve that in the future since now it works.      

c/c @estroz @DirectXMan12 could you help in the review?

**Comment by Adirio** [comment]
I just realized this has a +5% coverage 🎉

**Review by jmrodri — APPROVED** [review]
/lgtm

File: test/testdata/generate.sh [hunk 1]
```diff
@@ -96,13 +96,10 @@ function scaffold_test_project {
...
-    header_text 'enabling --pattern flag ...'
-    export KUBEBUILDER_ENABLE_PLUGINS=1
     header_text 'Creating APIs ...'
-    $kb create api --group crew --version v1 --kind Captain --controller=true --resource=true --make=false --pattern=addon
-    $kb create api --group crew --version v1 --kind FirstMate --controller=true --resource=true --make=false --pattern=addon
-    $kb create api --group crew --version v1 --kind Admiral --controller=true --resource=true --namespaced=false --make=false --pattern=addon
-    unset KUBEBUILDER_ENABLE_PLUGINS
+    $kb create api --group crew --version v1 --kind Captain --controller=true --resource=true --make=false
+    $kb create api --group crew --version v1 --kind FirstMate --controller=true --resource=true --make=false
+    $kb create api --group crew --version v1 --kind Admiral --controller=true --resource=true --namespaced=false --make=false
```
Comment (camilamacedo86) [inline test/testdata/generate.sh:102]: We have none scenario covering the `create api --plugins=declarative`. Could we add that? Maybe in the project-v3 and project-v2 WDYT

File: pkg/cli/options.go [hunk 2]
```diff
@@ -44,53 +44,47 @@ func WithVersion(version string) Option {
 	}
 }
 
-// WithDefaultProjectVersion is an Option that sets the cli's default project version.
-// Setting an unknown version will result in an error.
-func WithDefaultProjectVersion(version config.Version) Option {
+// WithPlugins is an Option that sets the cli's plugins.
```
Comment (camilamacedo86) [inline pkg/cli/options.go:47]: Could we improve all comments docs over what is exported with examples indeed of its usage?

File: test/testdata/generate.sh (hunk 1 above)
Comment (Adirio) [reply test/testdata/generate.sh:106]: Done, lets check if the changes in the corresponding testdata folder are sound.

File: test/testdata/generate.sh [hunk 3]
```diff
@@ -54,7 +54,11 @@ function scaffold_test_project {
       $kb create webhook --group crew --version v1 --kind Captain --defaulting --programmatic-validation --force
     fi
 
-    $kb create api --group crew --version v1 --kind FirstMate --controller=true --resource=true --make=false
+    if [ $project == "project-v2" ]; then
+      $kb create api --plugins="go/v2,declarative" --group crew --version v1 --kind FirstMate --controller=true --resource=true --make=false
+    else
+      $kb create api --plugins="go/v3,declarative" --group crew --version v1 --kind FirstMate --controller=true --resource=true --make=false
+    fi
```
Comment (camilamacedo86) [inline test/testdata/generate.sh:61]: 👍

File: testdata/project-v3-config/channels/stable [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Versions for the stable channel
+manifests:
+- version: 0.0.1
```
Comment (camilamacedo86) [inline testdata/project-v3-config/channels/stable:3]: Why we are doing this scaffold now?

File: testdata/project-v3/go.mod [hunk 5]
```diff
@@ -10,4 +10,5 @@ require (
 	k8s.io/apimachinery v0.19.2
 	k8s.io/client-go v0.19.2
 	sigs.k8s.io/controller-runtime v0.7.0
+	sigs.k8s.io/kubebuilder-declarative-pattern v0.0.0-20210113160450-b84d99da0217
```
Comment (camilamacedo86) [inline testdata/project-v3/go.mod:13]: @Adirio,

The project file is missing the info over the --plugin=declarative used to create the api for project-v2 and project-v3.

**Review by camilamacedo86 — COMMENTED** [review]
Hi @Adirio,
 
I am very happy with all changes here. 🥇  I also tested them locally and I have only 1 observation. I do not think that we can close #1941 with this one. See: https://github.com/kubernetes-sigs/kubebuilder/pull/2060/files#r588914741

So, could you only please comment on the generate testdata and update the first comment accordingly? 

Otherwise, it has my OK as well for we move forward. 

/approved

File: test/testdata/generate.sh (hunk 3 above)
Comment (camilamacedo86) [inline test/testdata/generate.sh:61]:
```suggestion
    # todo: improve the plugins api to address #1941 
    # if [ $project == "project-v2" ]; then
    #  $kb create api --plugins="go/v2,declarative" --group crew --version v1 --kind FirstMate --controller=true --resource=true --make=false
    # else
    #  $kb create api --plugins="go/v3,declarative" --group crew --version v1 --kind FirstMate --controller=true --resource=true --make=false
    # fi
```

To close #1941 it is missing: 
- By informing --plugins to the subcommand it is missing track the info in the PROJECT file.
- Add a help text for declarative plugin to allow us to check how the help text will be with `$ kubebuilder create api --plugins="go/v3,declarative"`
- shows the --plugins option for all subcommands with the cli help
- Improve the plugin subcommand overwrite behaviour to work as we spoke out in the issue #1941 project layout + values informed on the flag. See the motivations clarifies in the issue.

**Review by camilamacedo86 — APPROVED** [review]
Hi @Adirio,
 
I am very happy with all changes here. 🥇  I also tested them locally and I have only 1 observation. I do not think that we can close #1941 with this one. See: https://github.com/kubernetes-sigs/kubebuilder/pull/2060/files#r588914741

Otherwise, it has my OK as well for we move forward.

**Comment by camilamacedo86** [comment]
@Adirio 

This pr will also closes : https://github.com/kubernetes-sigs/kubebuilder/issues/1543

File: test/testdata/generate.sh (hunk 3 above)
Comment (Adirio) [reply test/testdata/generate.sh:61]: > * By informing --plugins to the subcommand it is missing track the info in the PROJECT file.

`declarative` is not tracking any info as you mention, will add a list of resources that were scaffolded with this plugin.

> * Add a help text for declarative plugin to allow us to check how the help text will be with `$ kubebuilder create api --plugins="go/v3,declarative"`

Will do

> * shows the --plugins option for all subcommands with the cli help

This is already working:
```
$ kubebuilder create api --plugins go/v3,declarative --help
[...]
Global Flags:
      --plugins strings   plugin keys to be used for this subcommand execution
```

File: testdata/project-v3-config/channels/stable (hunk 4 above)
Comment (Adirio) [reply testdata/project-v3-config/channels/stable:3]: That is part of the declarative pattern, it modifies `*_types.go` and `*_controller.go` and creates two placeholder files.

File: testdata/project-v3/go.mod (hunk 5 above)
Comment (Adirio) [reply testdata/project-v3/go.mod:13]: Done

File: test/testdata/generate.sh (hunk 3 above)
Comment (Adirio) [reply test/testdata/generate.sh:61]: Done

File: testdata/project-v3-config/PROJECT [hunk 6]
```diff
@@ -1,6 +1,13 @@
 componentConfig: true
 domain: testproject.org
 layout: go.kubebuilder.io/v3
+plugins:
```
Comment (camilamacedo86) [inline testdata/project-v3-config/PROJECT:4]: That is great :-)

**Comment by camilamacedo86** [comment]
Hi @Adirio, 

You are terrific. My concerns raised above are all addressed and it also closes #1941. 🥇

**Comment by pwittrock** [comment]
Thanks for doing this.  Looks good from a high level.  Will rely on others to look more closely at the code.

LGTM

/approve

**Comment by Adirio** [comment]
All pre-requisites have been merged, not canceling hold unitl #2075 gets reviewed and merged as it adds a couple new concepts found during the alignment with SDK.

File: pkg/cli/cmd_helpers.go [hunk 7]
```diff
@@ -17,69 +17,275 @@ limitations under the License.
...
+}
+
+// executionMethodsFuncs returns cobra RunE functions for PreRunE, RunE, PostRunE cobra hooks.
+func (c CLI) executionMethodsFuncs(
+	pluginKeys []string,
+	subcommands *[]plugin.Subcommand,
+	options *resourceOptions,
+	msg string,
+) (
+	func(*cobra.Command, []string) error,
```
Comment (DirectXMan12) [inline pkg/cli/cmd_helpers.go:135]: nit: return a struct or something for this -- more than two return values is pushing it in terms of readabilty

File: pkg/cli/cmd_helpers.go [hunk 8]
```diff
@@ -17,69 +17,275 @@ limitations under the License.
...
+			if err := res.Validate(); err != nil {
+				return fmt.Errorf("%s: created invalid resource: %w", msg, err)
+			}
+		}
+
+		// Pre-scaffold method.
+		subcommandsCopy = make([]plugin.Subcommand, len(*subcommands))
+		copy(subcommandsCopy, *subcommands)
+		for i, subcommand := range subcommandsCopy {
+			if subcmd, hasPreScaffold := subcommand.(plugin.HasPreScaffold); hasPreScaffold {
```
Comment (DirectXMan12) [inline pkg/cli/cmd_helpers.go:214]: might want to consider a helper for this pattern, considering it's repeated multiple times across this file -- could abstract over the error handling that skips certain subcommands/copying/etc, and just have something like

```go
err := ProcessSubcommands(subcommands, func(cmd Subcommand) error {
  subcmd, hasPreScaffold := cmd.HasPreScaffold
  if !hasPreScaffold {
    return nil
  }
  return subcmd.PreScaffold(fs)
})
```

File: pkg/cli/edit.go [hunk 9]
```diff
@@ -14,77 +14,58 @@ See the License for the specific language governing permissions and
...
+		cmdErr(cmd, noAvailablePluginError{"edit project"})
+		return cmd
 	}
 
-	subcommand := editPlugin.GetEditSubcommand()
-	subcommand.BindFlags(cmd.Flags())
-	subcommand.UpdateContext(&ctx)
-	cmd.Long = ctx.Description
-	cmd.Example = ctx.Examples
+	// Initialization methods.
```
Comment (DirectXMan12) [inline pkg/cli/edit.go:64]: ?? this comment doesn't actually say much

File: pkg/cli/cmd_helpers.go [hunk 10]
```diff
@@ -17,69 +17,275 @@ limitations under the License.
...
+	for _, p := range plugins {
+		if filter(p) {
+			pluginKeys = append(pluginKeys, plugin.KeyFor(p))
+			subcommands = append(subcommands, extract(p))
+		}
+	}
+	return pluginKeys, &subcommands
+}
+
+// initializationMethods
```
Comment (DirectXMan12) [inline pkg/cli/cmd_helpers.go:91]: this needs better godoc

**Review by DirectXMan12 — COMMENTED** [review]
couple of nits.  Main comment is there's this pattern of iterating over the commands, checking if they match and interface, trying to do something if they do, and skipping them if that fails.  We should have a helper for that -- I think it'd make the code a lot clearer

File: pkg/cli/cmd_helpers.go [hunk 11]
```diff
@@ -17,69 +17,331 @@ limitations under the License.
...
+	errorMessage string
+	// projectVersion is the project version that will be used to create new project configurations.
+	// It is only used for initialization.
+	projectVersion config.Version
+	// layout is the plugin chain configured for this project.
+	layout string
+}
+
+// subcommandTuples iterates over the subcommands skipping those flagged to do so.
+func (factory *executionHooksFactory) subcommandTuples() chan keySubcommandTuple {
```
Comment (DirectXMan12) [inline pkg/cli/cmd_helpers.go:198]: I don't want to make this too absurd, but is there any reason we can't simplify these three methods down to

```go
func newFilter(subcmds []plugin.Subcommand) {
  keys := /* extract keys */
  return &filteredSubcommands {
    subcommands: subcmds,
    keys: keys,
    skip: make([]bool, len(subcmds)),
  }
}
type filteredSubcommands struct {
  subcommands []plugin.Subcommand
  keys []string
  skip []bool
}
func (s *filteredSubcommands) ForEach(cb func(subcmd plugin.Subcommand) error) error {
  for i, cmd := range s.subcommands {
    if s.skip[i] {
        continue
    }
    
    err := cb(cmd)
    if err != nil {
        var exitError plugin.ExitError
        if errors.As(err, &exitError) {
          s.skip[i] = true
          fmt.Printf("skipping remaining hooks for %q: %s\n", s.keys[i], exitError.Reason)
          continue
        }
        return fmt.Errorf("for %s, %w", s.keys[i], err)
    }
  }
  return nil
}
````
 
Then below, using it becomes

```go
filterer := newFilter(subcommands)

// ...

err := filterer.ForEach(func(subcmd plugin.Subcommand) error {
  if subcmd, reqCfg := subcmd.(plugin.RequiresConfig); reqCfg {
    return subcmd.InjectConfig(cfg)
  }
  return nil
})
if err != nil {
  return fmt.Errorf("unable to inject configuration: %w", err)
}

// etc
```

**Comment by DirectXMan12** [comment]
/lgtm

can clean up later if we need to

File: pkg/cli/cmd_helpers.go (hunk 11 above)
Comment (Adirio) [reply pkg/cli/cmd_helpers.go:198]: Will apply this approach if I don't see any drawback and rebase

File: pkg/cli/cli.go [hunk 12]
```diff
@@ -157,258 +137,248 @@ func newCLI(options ...Option) (*CLI, error) {
...
 }
 
 // getInfoFromConfig obtains the project version and plugin keys from the project config.
 // It is extracted from getInfoFromConfigFile for testing purposes.
-func getInfoFromConfig(projectConfig config.Config) (config.Version, []string, error) {
+func (c *CLI) getInfoFromConfig(projectConfig config.Config) error {
 	// Split the comma-separated plugins
 	var pluginSet []string
 	if projectConfig.GetLayout() != "" {
 		for _, p := range strings.Split(projectConfig.GetLayout(), ",") {
```
[the remark this answers is not in the document]
Comment (Adirio) [reply pkg/cli/cli.go:203]: Solved in #2096

File: pkg/cli/cmd_helpers.go (hunk 11 above)
Comment (Adirio) [reply pkg/cli/cmd_helpers.go]: Done

**Comment by camilamacedo86** [comment]
All requested changes were done. 
It has the approval of @pwittrock, @jmrodri, @DirectXMan12 and me
The PR in SDK as this one has been reviewed by @joelanford, @estroz and me 
As @DirectXMan12 described we always can improve things and refractory the stuff latter as well. 

It introduces a good value and is a blocker for we move forward with the other ideas such as https://github.com/kubernetes-sigs/kubebuilder/issues/2015, https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/code-generate-image-plugin.md 

So, IMO we can move forward with and get it merged
It has my 

/lgtm 

too.

**Pull request merged**